File: Proton.cpp
#include "Proton.h"
#include "BatchRenderer.h"
#include "AtomManager.h"
#include <cmath>
#include <algorithm>

Proton::Proton(sf::Vector2f position, sf::Vector2f velocity, sf::Color color, float energy)
    : m_position(position)
    , m_velocity(velocity)
    , m_color(color)
    , m_energy(energy)
    , m_isAlive(true)
    , m_markedForDeletion(false)
    , m_lifetime(0.0f)
    , m_pulseTimer(0.0f)
    , m_charge(+1)
    , m_neutronCount(0)
    , m_isStableHydrogen(false)
    , m_waveFieldTimer(0.0f)
{
    m_radius = calculateRadius(energy);
    m_mass = calculateMass(energy);

    // Proton lifetime
    m_maxLifetime = Constants::Proton::DEFAULT_LIFETIME;
    m_fadeStartTime = m_maxLifetime * Constants::Proton::FADE_START_RATIO;
}

void Proton::update(float deltaTime, const sf::Vector2u& windowSize)
{
    if (!m_isAlive) return;

    // Update lifetime
    m_lifetime += deltaTime;
    m_pulseTimer += deltaTime;

    // Check if proton should die from age (if maxLifetime >= 0)
    if (m_maxLifetime >= 0 && m_lifetime >= m_maxLifetime)
    {
        m_isAlive = false;
        return;
    }

    // Friction removed to simulate vacuum
    // m_velocity *= FRICTION;

    // Update position based on velocity
    m_position += m_velocity * deltaTime;

    // Handle boundary collisions
    handleBoundaryCollision(windowSize);
}

void Proton::addToBatch(BatchRenderer& batchRenderer) const
{
    if (!m_isAlive) return;

    // Calculate visual properties based on lifetime
    sf::Color renderColor = m_color;
    float renderRadius = m_radius;

    // Apply charge state visual feedback
    if (m_isStableHydrogen)
    {
        // Stable hydrogen: bright white
        renderColor.r = Constants::Proton::STABLE_HYDROGEN_R;
        renderColor.g = Constants::Proton::STABLE_HYDROGEN_G;
        renderColor.b = Constants::Proton::STABLE_HYDROGEN_B;
        renderRadius *= Constants::Proton::STABLE_HYDROGEN_RADIUS_MULTIPLIER;
    }
    else if (m_charge == 0)
    {
        // Neutral with neutron: gray-white
        renderColor.r = Constants::Proton::NEUTRAL_PROTON_R;
        renderColor.g = Constants::Proton::NEUTRAL_PROTON_G;
        renderColor.b = Constants::Proton::NEUTRAL_PROTON_B;
    }
    else if (m_charge == +1)
    {
        // Bare proton: slight red tint
        renderColor.r = static_cast<std::uint8_t>(std::min(255, static_cast<int>(renderColor.r * Constants::Proton::BARE_PROTON_RED_TINT)));
    }

    // Pulsing effect based on energy
    float pulseFrequency = Constants::Proton::PULSE_FREQUENCY_BASE + (m_energy * Constants::Proton::PULSE_FREQUENCY_ENERGY_FACTOR);
    float pulse = std::sin(m_pulseTimer * pulseFrequency) * Constants::Proton::PULSE_INTENSITY + Constants::Proton::PULSE_BASE;
    renderRadius *= pulse;

    // Fade out near end of lifetime
    if (m_maxLifetime >= 0 && m_lifetime > m_fadeStartTime)
    {
        float fadeRatio = (m_lifetime - m_fadeStartTime) / (m_maxLifetime - m_fadeStartTime);
        float fadeAmount = 1.0f - fadeRatio;
        renderColor.a = static_cast<std::uint8_t>(255 * fadeAmount);
    }

    // Render core proton
    batchRenderer.addAtom(m_position, renderRadius, renderColor);

    // Add glow layers for visual polish
    sf::Color glowColor1 = renderColor;
    glowColor1.a = static_cast<std::uint8_t>(glowColor1.a * Constants::Proton::GLOW_LAYER1_ALPHA);
    batchRenderer.addAtom(m_position, renderRadius * Constants::Proton::GLOW_LAYER1_RADIUS, glowColor1);

    sf::Color glowColor2 = renderColor;
    glowColor2.a = static_cast<std::uint8_t>(glowColor2.a * Constants::Proton::GLOW_LAYER2_ALPHA);
    batchRenderer.addAtom(m_position, renderRadius * Constants::Proton::GLOW_LAYER2_RADIUS, glowColor2);
}

void Proton::absorbProton(const Proton& other)
{
    // Combine energy
    float totalEnergy = m_energy + other.m_energy;

    // Combine momentum (mass * velocity)
    float totalMass = m_mass + other.m_mass;
    sf::Vector2f totalMomentum = m_velocity * m_mass + other.m_velocity * other.m_mass;

    // Update properties
    m_energy = totalEnergy;
    m_radius = calculateRadius(m_energy);
    m_mass = calculateMass(m_energy);
    m_velocity = totalMomentum / totalMass;

    // Mix colors (weighted by energy)
    float weight1 = m_energy / totalEnergy;
    float weight2 = other.m_energy / totalEnergy;

    m_color.r = static_cast<std::uint8_t>(m_color.r * weight1 + other.m_color.r * weight2);
    m_color.g = static_cast<std::uint8_t>(m_color.g * weight1 + other.m_color.g * weight2);
    m_color.b = static_cast<std::uint8_t>(m_color.b * weight1 + other.m_color.b * weight2);
}

float Proton::calculateRadius(float energy) const
{
    // Scale radius based on energy, clamped to min/max
    float radius = Constants::Proton::MIN_RADIUS + (energy * Constants::Proton::ENERGY_TO_RADIUS_FACTOR);
    return std::clamp(radius, Constants::Proton::MIN_RADIUS, Constants::Proton::MAX_RADIUS);
}

float Proton::calculateMass(float energy) const
{
    // Mass proportional to energy (E=mc²-ish)
    return energy * Constants::Proton::ENERGY_TO_MASS_FACTOR;
}

void Proton::handleBoundaryCollision(const sf::Vector2u& windowSize)
{
    bool collided = false;

    // Left/right boundaries
    if (m_position.x - m_radius < 0)
    {
        m_position.x = m_radius;
        m_velocity.x = -m_velocity.x * Constants::Proton::BOUNCE_DAMPENING;
        collided = true;
    }
    else if (m_position.x + m_radius > windowSize.x)
    {
        m_position.x = windowSize.x - m_radius;
        m_velocity.x = -m_velocity.x * Constants::Proton::BOUNCE_DAMPENING;
        collided = true;
    }

    // Top/bottom boundaries
    if (m_position.y - m_radius < 0)
    {
        m_position.y = m_radius;
        m_velocity.y = -m_velocity.y * Constants::Proton::BOUNCE_DAMPENING;
        collided = true;
    }
    else if (m_position.y + m_radius > windowSize.y)
    {
        m_position.y = windowSize.y - m_radius;
        m_velocity.y = -m_velocity.y * Constants::Proton::BOUNCE_DAMPENING;
        collided = true;
    }
}

void Proton::tryNeutronFormation(float deltaTime, bool nearAtom)
{
    // Already has neutron, skip
    if (m_charge != +1) return;

    // Not near atom, reset timer
    if (!nearAtom)
    {
        m_waveFieldTimer = 0.0f;
        return;
    }

    // Near atom, accumulate time
    m_waveFieldTimer += deltaTime;

    // Check if neutron formation threshold reached
    if (m_waveFieldTimer >= Constants::Proton::NEUTRON_FORMATION_TIME)
    {
        m_neutronCount = 1;
        m_charge = 0;
        m_radius *= Constants::Proton::NEUTRON_RADIUS_MULTIPLIER;
        m_waveFieldTimer = 0.0f;
    }
}

bool Proton::tryCaptureElectron(const PathFollowingAtom& electron)
{
    // Need to have neutral charge (needs neutron first)
    if (m_charge != 0) return false;

    // Need to have exactly 1 neutron
    if (m_neutronCount != 1) return false;

    // Already stable, skip
    if (m_isStableHydrogen) return false;

    // Calculate distance to electron
    sf::Vector2f delta = electron.getPosition() - m_position;
    float distance = std::sqrt(delta.x * delta.x + delta.y * delta.y);

    // Check if electron is close enough to capture
    if (distance < Constants::Proton::ELECTRON_CAPTURE_DISTANCE)
    {
        m_isStableHydrogen = true;
        m_maxLifetime = Constants::Proton::INFINITE_LIFETIME; // Never die from age
        return true;
    }

    return false;
}

File: RustPond/src/proton.rs
// Proton - Direct port from Proton.h/cpp
// Rare, persistent physics particle with nuclear fusion capabilities

use macroquad::prelude::*;
use crate::constants::*;
use crate::constants::proton as pc;

#[derive(Clone)]
pub struct Proton {
    position: Vec2,
    velocity: Vec2,
    color: Color,
    energy: f32,
    radius: f32,
    mass: f32,
    is_alive: bool,
    marked_for_deletion: bool,
    lifetime: f32,
    max_lifetime: f32,

    // Visual effects
    pulse_timer: f32,
    fade_start_time: f32,

    // Charge state system
    charge: i32,
    neutron_count: i32,
    is_stable_hydrogen: bool,
    wave_field_timer: f32,

    // Sleeping system for optimization
    is_sleeping: bool,
}

impl Proton {
    pub fn new(position: Vec2, velocity: Vec2, color: Color, energy: f32, charge: i32) -> Self {
        let radius = Self::calculate_radius(energy);
        let mass = Self::calculate_mass(energy);
        let max_lifetime = pc::DEFAULT_LIFETIME;
        let fade_start_time = max_lifetime * pc::FADE_START_RATIO;

        Self {
            position,
            velocity,
            color,
            energy,
            radius,
            mass,
            is_alive: true,
            marked_for_deletion: false,
            lifetime: 0.0,
            max_lifetime,
            pulse_timer: 0.0,
            fade_start_time,
            charge,
            neutron_count: 0,
            is_stable_hydrogen: false,
            wave_field_timer: 0.0,
            is_sleeping: false,
        }
    }

    pub fn update(&mut self, delta_time: f32, window_size: (f32, f32)) {
        if !self.is_alive {
            return;
        }

        // Always update visual pulse
        self.pulse_timer += delta_time;

        // SLEEPING OPTIMIZATION
        if self.is_stable_hydrogen || self.is_stable_helium4() {
            let velocity_magnitude = self.velocity.length();

            if velocity_magnitude < 1.0 {
                self.is_sleeping = true;
                return; // Skip all physics
            } else {
                self.is_sleeping = false;
            }
        }

        // Update lifetime
        self.lifetime += delta_time;

        // Check death from age
        if self.max_lifetime >= 0.0 && self.lifetime >= self.max_lifetime {
            self.is_alive = false;
            return;
        }

        // Straight-line movement
        self.position += self.velocity * delta_time;

        // Boundary collisions
        self.handle_boundary_collision(window_size);

        // Off-screen culling
        const CULL_MARGIN: f32 = 200.0;
        if self.position.x < -CULL_MARGIN || self.position.x > window_size.0 + CULL_MARGIN ||
           self.position.y < -CULL_MARGIN || self.position.y > window_size.1 + CULL_MARGIN {
            if !self.is_stable_hydrogen && self.charge != 2 {
                self.is_alive = false;
            }
        }
    }

    fn handle_boundary_collision(&mut self, window_size: (f32, f32)) {
        let mut collided = false;

        // Left/right
        if self.position.x - self.radius < 0.0 {
            self.position.x = self.radius;
            self.velocity.x = -self.velocity.x * pc::BOUNCE_DAMPENING;
            collided = true;
        } else if self.position.x + self.radius > window_size.0 {
            self.position.x = window_size.0 - self.radius;
            self.velocity.x = -self.velocity.x * pc::BOUNCE_DAMPENING;
            collided = true;
        }

        // Top/bottom
        if self.position.y - self.radius < 0.0 {
            self.position.y = self.radius;
            self.velocity.y = -self.velocity.y * pc::BOUNCE_DAMPENING;
            collided = true;
        } else if self.position.y + self.radius > window_size.1 {
            self.position.y = window_size.1 - self.radius;
            self.velocity.y = -self.velocity.y * pc::BOUNCE_DAMPENING;
            collided = true;
        }

        if collided {
            self.is_sleeping = false;
        }
    }

    pub fn try_neutron_formation(&mut self, delta_time: f32, near_atom: bool) {
        if self.charge != 1 {
            return;
        }

        if !near_atom {
            self.wave_field_timer = 0.0;
            return;
        }

        self.wave_field_timer += delta_time;

        if self.wave_field_timer >= pc::NEUTRON_FORMATION_TIME {
            self.neutron_count = 1;
            self.charge = 0;
            self.radius *= pc::NEUTRON_RADIUS_MULTIPLIER;
            self.wave_field_timer = 0.0;
        }
    }

    pub fn try_capture_electron(&mut self, electron_pos: Vec2) -> bool {
        if self.charge != 0 || self.neutron_count != 1 || self.is_stable_hydrogen {
            return false;
        }

        let distance = self.position.distance(electron_pos);

        if distance < pc::ELECTRON_CAPTURE_DISTANCE {
            self.is_stable_hydrogen = true;
            self.max_lifetime = pc::INFINITE_LIFETIME;
            return true;
        }

        false
    }

    pub fn get_element_label(&self) -> String {
        if self.charge == 2 && self.neutron_count == 2 {
            "He4".to_string()
        } else if self.charge == 1 && self.neutron_count == 2 {
            "He3".to_string()
        } else if self.charge == -1 {
            "H-".to_string()
        } else if self.charge == 0 && self.neutron_count == 1 {
            "H".to_string()
        } else if self.charge == 1 && self.neutron_count == 0 {
            "H+".to_string()
        } else if self.is_stable_hydrogen {
            "H1".to_string()
        } else {
            "?".to_string()
        }
    }

    pub fn render(&self, segments: i32) {
        if !self.is_alive {
            return;
        }

        let mut render_color = self.color;
        let mut render_radius = self.radius;

        // Apply charge state visuals
        if self.is_stable_hydrogen {
            render_color = Color::from_rgba(255, 255, 255, 255);
            render_radius *= pc::STABLE_HYDROGEN_RADIUS_MULTIPLIER;
        } else if self.charge == 0 {
            render_color = Color::from_rgba(200, 200, 200, 255);
        } else if self.charge == 1 {
            let r = (render_color.r * pc::BARE_PROTON_RED_TINT).min(1.0);
            render_color.r = r;
        } else if self.charge == -1 {
            let b = (render_color.b * 1.3).min(1.0);
            render_color.b = b;
        }

        // Helium-3
        if self.charge == 1 && self.neutron_count == 2 {
            render_color = Color::from_rgba(255, 200, 100, 255);
            render_radius *= pc::HELIUM3_RADIUS_MULTIPLIER;
        }
        // Helium-4
        else if self.charge == 2 && self.neutron_count == 2 {
            render_color = Color::from_rgba(255, 255, 100, 255);
            render_radius *= pc::HELIUM4_RADIUS_MULTIPLIER;
        }

        // Pulsing effect
        let pulse_frequency = pc::PULSE_FREQUENCY_BASE + (self.energy * pc::PULSE_FREQUENCY_ENERGY_FACTOR);
        let pulse = (self.pulse_timer * pulse_frequency).sin() * pc::PULSE_INTENSITY + pc::PULSE_BASE;
        render_radius *= pulse;

        // Fade out near end of lifetime
        if self.max_lifetime >= 0.0 && self.lifetime > self.fade_start_time {
            let fade_ratio = (self.lifetime - self.fade_start_time) / (self.max_lifetime - self.fade_start_time);
            let fade_amount = 1.0 - fade_ratio;
            render_color.a = fade_amount;
        }

        // Draw core
        draw_poly(self.position.x, self.position.y, segments as u8, render_radius, 0.0, render_color);

        // Glow layer 1
        let mut glow1 = render_color;
        glow1.a *= pc::GLOW_LAYER1_ALPHA;
        draw_poly(self.position.x, self.position.y, segments as u8, render_radius * pc::GLOW_LAYER1_RADIUS, 0.0, glow1);

        // Glow layer 2
        let mut glow2 = render_color;
        glow2.a *= pc::GLOW_LAYER2_ALPHA;
        draw_poly(self.position.x, self.position.y, segments as u8, render_radius * pc::GLOW_LAYER2_RADIUS, 0.0, glow2);
    }

    fn calculate_radius(energy: f32) -> f32 {
        let radius = pc::MIN_RADIUS + (energy * pc::ENERGY_TO_RADIUS_FACTOR);
        radius.clamp(pc::MIN_RADIUS, pc::MAX_RADIUS)
    }

    fn calculate_mass(energy: f32) -> f32 {
        energy * pc::ENERGY_TO_MASS_FACTOR
    }

    // Getters
    pub fn is_alive(&self) -> bool { self.is_alive && !self.marked_for_deletion }
    pub fn is_marked_for_deletion(&self) -> bool { self.marked_for_deletion }
    pub fn position(&self) -> Vec2 { self.position }
    pub fn velocity(&self) -> Vec2 { self.velocity }
    pub fn radius(&self) -> f32 { self.radius }
    pub fn energy(&self) -> f32 { self.energy }
    pub fn mass(&self) -> f32 { self.mass }
    pub fn color(&self) -> Color { self.color }
    pub fn charge(&self) -> i32 { self.charge }
    pub fn neutron_count(&self) -> i32 { self.neutron_count }
    pub fn is_stable_hydrogen(&self) -> bool { self.is_stable_hydrogen }
    pub fn is_stable_helium4(&self) -> bool { self.charge == 2 && self.neutron_count == 2 }
    pub fn is_sleeping(&self) -> bool { self.is_sleeping }

    // Setters
    pub fn set_velocity(&mut self, velocity: Vec2) {
        self.velocity = velocity;
        self.is_sleeping = false;
    }
    pub fn mark_for_deletion(&mut self) { self.marked_for_deletion = true; }
    pub fn set_neutron_count(&mut self, count: i32) { self.neutron_count = count; }
    pub fn set_max_lifetime(&mut self, lifetime: f32) { self.max_lifetime = lifetime; }
    pub fn wake(&mut self) { self.is_sleeping = false; }
}
